pandas/io/json/_json.py

```diff
@@ -726,7 +742,13 @@ def read(self):
             obj = concat(self)
         elif self.lines:
             data = ensure_str(self.data)
-            obj = self._get_object_parser(self._combine_lines(data.split("\n")))
+            if self.nrows:
+                compiled_pattern = re.compile(".*\n")
+                data = (line.group(0) for line in compiled_pattern.finditer(data))
+                data = islice(data, self.nrows)
```

Comment:
ok you are nicely keeping an iterator up to this point.

But I *think* you are materializing it fully when it gets passed to on L751.

Instead what we need to do is more of this pattern: https://github.com/pandas-dev/pandas/blob/master/pandas/io/pytables.py#L1715

IOW what you have to do is create a JSONIterator object *always* , then you return it.get_result() if there is NO chunksize, else return the iterator (when there is a chunksize). The user can then iterator over this object and each time it materializes the chunk *only*

make sense?

Reply:
Thanks for the review.

Reply:
Hi @jreback did the changes 

If the nrows is set  StringIO is used instead of reading the entire file using data. This is how chunksize is handled too. 

Later in the read method,  islice method is used on the StringIO iterator to slice the string without loading the entire string in memory.

Thanks for the reviews

Reply:
this *still* reads the entire string into memory, that's the entire point of chunksize to avoid this. please use the above method that I pointed in combination with the finditer.

Reply:
Please check the new commit I made 6 hours.
I am doing the exact same things as done in chunksize __next__ method.

Reply:
Like  islice is done on StringIO in the __next__ method using chunksize.   I am doing  islice on the StringIO using the nrows parameter. I really do not think that we need the finditer iterator as we can use the same StringIO instance for both the case. and it is not loading the entire string in memory
`       if not hasattr(data, "read") and (self.chunksize or self.nrows):data = StringIO(data)`

This block makes sure that if either one of the chunksize or nrows are set it does not load the entire string in memory.

In the iterator's next method we do islice on the self.data 
In the same way I am applying islice on the self.data which is actually a StringIO object when self.nrows is set and chunksize is not set.

Reply:
ok can you show a benchmark (actually also add this to asv you can do a memory benchmark), make a big string (doesn't have to be huge) and time with with chunksize (give a couple like 1, 10, 100 or something) and nrows and then no parameters.

Reply:
Thanks for the quick response
I actually started doing a basic benchmark test.
If this works I will go ahead with adding it to the asv.

Reply:
Thanks for the quick response
I actually started doing a basic benchmark test.
I will add it to the asv.

Reply:
Hey @jreback ,

Thanks for the suggestions

I did the time benchmarks.
I believe the analysis shows that the entire string is not loaded at once when the nrows and chunk size are set.

The code is here [link](https://pastebin.com/Cb7h5jDa)
The output is here [link](https://pastebin.com/NMUKA65T)

Please do check and tell me if I missed something while writing the test

Reply:
Can you add an ASV benchmark? You can model it after these existing ones in the JSON benchmark module

https://github.com/pandas-dev/pandas/blob/cc634842a1d73e77145d62c4390dfa10f282e06b/asv_bench/benchmarks/io/json.py#L56

Reply:
yeah for these we could like memory & time benchmarks

Reply:
Hey,
Thanks for the review.
I haven't worked on asv benchmarks before. But I believe I should be able to learn and do it.

Reply:
Hey @jreback and @WillAyd.
I added the ASV benchmarks. I hope everything is in order.

Reply:
Can you post the results of running them here? Probably only need one of each instead of a 15k/45k variant but depending on how long they run we can decide which to keep

Reply:
Hey @WillAyd ,

I pasted it here
[Link](https://pastebin.com/BMpCyC4m)

Reply:
can u add chunk size benchmark as well

Reply:
Hey it was already there in will put the results in an another paste

[Link](https://pastebin.com/VwuDiXMy)

But I don't know how much useful this is. It actually benchmarks to get all the chunks

Reply:
it is useful - it’s showing that using chunk uses the same memory
hence it’s not working properly

chunk should be about equal to the memory of a single chunk of string plus the final dataframe size

Reply:
I believe the memory usage is that high because concat is applied on the generator.

So as per the implementation of read_json.when read_json is called with chunksize it returns an iterator to the user. 

As the benchmark is calling concat on that iterator. All the chunks are loaded into the memory one by one and each is concatenated to the same dataframe, which is returned.

So I believe the issue is with the benchmark and not with chunksize implementation.

Reply:
using chunksize should provide a lower memory usage

can u show using various chunk sizes like 1., 100, 10000 on the same dataset

Reply:
okay will add these benchmarks

Reply:
@jreback 
Added the benchmarks and have created this spreadsheet with the results to compare side by side
[spreadsheet](https://docs.google.com/spreadsheets/d/1MjE2iMrlYnWbnfSyizfYI3IbwSK1z760VKeoYfCBeJs/edit?usp=sharing)

Reply:
updated the spreadsheet with the latest results

Reply:
> Can you post the results of running them here? Probably only need one of each instead of a 15k/45k variant but depending on how long they run we can decide which to keep


Hey @WillAyd 
These are the results of running the 15k and 45k variants. I just wanted to ask you which one of them should we keep or maybe another number. Also, can you tell me how do we decide on this so further I would be able to decide this  on my own.


2020-05-30T11:18:23.3074450Z [ 25.32%] ··· ...son.ReadJSONLines.time_read_json_lines_nrows                 ok
2020-05-30T11:18:23.3074773Z [ 25.32%] ··· ========== ==========
2020-05-30T11:18:23.3074960Z                 index              
2020-05-30T11:18:23.3075257Z               ---------- ----------
2020-05-30T11:18:52.9832677Z                  int      83.3±0ms
2020-05-30T11:18:52.9833739Z                datetime   77.6±0ms
2020-05-30T11:18:52.9834220Z               ========== ==========
2020-05-30T11:18:52.9834514Z
2020-05-30T11:18:52.9835186Z [ 25.38%] ··· ...dJSONLines.time_read_json_lines_nrows_larger                 ok
2020-05-30T11:18:52.9835777Z [ 25.38%] ··· ========== =========
2020-05-30T11:18:52.9836148Z                 index            
2020-05-30T11:18:52.9838624Z               ---------- ---------
2020-05-30T11:18:52.9839011Z                  int      179±0ms
2020-05-30T11:18:52.9839288Z                datetime   175±0ms
2020-05-30T11:18:52.9839411Z               ========== =========


The entire related dump is [here ](https://pastebin.com/E22PAxKe) if you wanna have a look